src/main/scala/org/apache/spark/shuffle/ucx/UcxShuffleTransport.scala

```diff
@@ -0,0 +1,138 @@
+/*
+* Copyright (c) 2020, NVIDIA CORPORATION. ALL RIGHTS RESERVED.
+* See file LICENSE for terms.
+*/
+package org.apache.spark.shuffle.ucx
+
+/**
+ * Class that represents some block in memory with it's address, size.
+ * @param isHostMemory - host or GPU memory
+ */
+case class UcxMemoryBlock(address: Long, size: Long, isHostMemory: Boolean = true)
+
+/**
+ * Opaque object to describe remote memory block (address, rkey, etc.).
+ */
+trait Cookie {
+  // Write this cookie to some address in memory
+  def writeToMemory(memory: UcxMemoryBlock)
+
+  // Size of this cookie in bytes
+  def size: Int
+}
+
+/**
+ * Base class to indicate some blockId. It should be hashable and could be constructed on both ends.
+ * E.g. ShuffleBlockId(shuffleId: Int, mapId: Long, reduceId: Int)
+ */
+trait BlockId
+
+/**
+ * Special case for metadata blocks, that always would be fetched using
```

Comment:
Lets not have metadata in the api at all as discussed.

Instead of doing this, if you really want a hint from the application that a data block is needed, please add a `prefetchByCookie`: let the application decide when the data blocks should be prefetched.